src/algebra/add_torsor.lean

```diff
@@ -0,0 +1,191 @@
+/-
+Copyright (c) 2020 Joseph Myers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Joseph Myers.
+-/
+import algebra.group.basic
+
+/-!
+# Torsors of additive group actions
+
+This file defines torsors of additive group actions.
+
+## Notations
+
+The group elements are referred to as acting on points.  This file
+defines the notation `+ᵥ` for adding a group element to a point and
+`-ᵥ` for subtracting two points to produce a group element.
+
+## Implementation notes
+
+Affine spaces are the motivating example of torsors of additive group
+actions.  It may be appropriate to refactor in terms of the general
+definition of group actions, via `to_additive`, when there is a use
+for multiplicative torsors (currently mathlib only develops the theory
+of group actions for multiplicative group actions).  The variable `G`
+is an explicit rather than implicit argument to lemmas because
+otherwise the elaborator sometimes has problems inferring appropriate
+types and type class instances.
+
+## References
+
+* https://en.wikipedia.org/wiki/Principal_homogeneous_space
+* https://en.wikipedia.org/wiki/Affine_space
+
+-/
+
+/-- Type class for the `+ᵥ` notation. -/
+class has_vadd (G : Type*) (P : Type*) :=
+(vadd : G → P → P)
+
+/-- Type class for the `-ᵥ` notation. -/
+class has_vsub (G : Type*) (P : Type*) :=
+(vsub : P → P → G)
+
+infix ` +ᵥ `:65 := has_vadd.vadd
+infix ` -ᵥ `:65 := has_vsub.vsub
+
+section prio
+set_option default_priority 100 -- see Note [default priority]
+set_option old_structure_cmd true
+/-- Type class for additive monoid actions. -/
+class add_action (G : Type*) (P : Type*) [add_monoid G] extends has_vadd G P :=
+(zero_vadd' : ∀ p : P, (0 : G) +ᵥ p = p)
+(vadd_assoc' : ∀ (g1 g2 : G) (p : P), g1 +ᵥ (g2 +ᵥ p) = (g1 + g2) +ᵥ p)
+
+/-- An `add_torsor G P` gives a structure to the nonempty type `P`,
+acted on by an `add_group G` with a transitive and free action given
+by the `+ᵥ` operation and a corresponding subtraction given by the
+`-ᵥ` operation. In the case of a vector space, it is an affine
+space. -/
+class add_torsor (G : Type*) (P : Type*) [add_group G] extends add_action G P, has_vsub G P :=
+(nonempty : nonempty P)
```

Comment:
Should this be `[nonempty : nonempty P]`?

Reply:
I don't know. What are the merits of one approach versus the other, once it's a field of the structure as requested above?

Reply:
With `[nonempty : nonempty P]` it'll try to get this argument from instance search instead of asking a user to write `nonempty := by apply_instance`.

Reply:
Thanks, I've made that change (and thus, having merged in master to get the required instance, eliminated the `nonempty := ⟨0⟩` in the instance of a group as a torsor for itself).